### PR TITLE
Enhance SimpleToastWidget with Close Button 

### LINKED ISSUE #106: Add ability to disable title
It's so nice to disable title. Showing only body is more simple and smart.
Additionally, disabling icon is also good.

### PR DESCRIPTION
**Pull request by AntonioAEMartins**
**Key Features:**

1. Close Button Support:

The SimpleToastWidget now includes a close button that can be configured using the showCloseButton argument. This allows the widget to:

- Always show the close button.

- Hide the close button.

- Support custom behaviors where the close button can be conditionally rendered (e.g., when integrated with styles that provide hover behavior).

2. Customizable Close Button Behavior:

By utilizing the ToastificationStyle, users can decide if the SimpleToastWidget should:

- Show a permanent close button.

- Omit the close button entirely.

- Customize for advanced behaviors (like on hover, if implemented externally).

close #106

### CODE REVIEW

**Comment by AntonioAEMartins** [comment]
Hi @payam-zahedi,

I hope you're doing well. As per our discussion on issue #106, I’ve implemented the close button functionality for the SimpleToastWidget. I’ve tested it using the Toastification WebApp Page, and everything is functioning smoothly.

One area I wanted to highlight is related to the onHover button type. I aimed to implement an animation for this case to handle the expansion of the widget when the button appears on hover. Initially, I faced a challenge where, if the button space is pre-allocated, the text isn’t centered properly. On the other hand, if I don’t pre-allocate the space, the toast expands when hovered.

Please let me know if there are any changes or improvements you'd suggest. I’m open to refining it further based on your recommendations.

**Comment by payam-zahedi** [comment]
Hey @AntonioAEMartins 

Thanks for your contribution

regarding on hover and toast allocated space I should try it to see what is the best practice 

I will check it and let you know

File: lib/src/widget/built_in/built_in_builder.dart [hunk 1]
```diff
@@ -344,6 +344,7 @@ class BuiltInToastBuilder extends StatelessWidget {
               boxShadow: boxShadow,
               direction: direction,
               applyBlurEffect: applyBlurEffect ?? false,
+              showCloseButton: showCloseButton,
             ),
```
Comment (payam-zahedi) [inline lib/src/widget/built_in/built_in_builder.dart:348]: You need to  pass `onCloseTap` parameter as well

File: lib/src/widget/built_in/simple/simple.dart [hunk 2]
```diff
@@ -44,28 +46,42 @@ class SimpleToastWidget extends StatelessWidget {
...
 
+  final bool? showCloseButton;
+  final VoidCallback? onCloseTap;
+
   SimpleStyle get defaultStyle => SimpleStyle(type);
 
   @override
-  Widget build(BuildContext context) {
-    final background = backgroundColor ?? defaultStyle.backgroundColor(context);
+  _SimpleToastWidgetState createState() => _SimpleToastWidgetState();
```
Comment (payam-zahedi) [inline lib/src/widget/built_in/simple/simple.dart:55]: We don't need to make this widget as Statefull we don't have any state in this widget

File: lib/src/widget/built_in/simple/simple.dart [hunk 3]
```diff
@@ -44,28 +46,42 @@ class SimpleToastWidget extends StatelessWidget {
...
+}
 
+class _SimpleToastWidgetState extends State<SimpleToastWidget>
+    with SingleTickerProviderStateMixin {
```
Comment (payam-zahedi) [inline lib/src/widget/built_in/simple/simple.dart:59]: We don't need SingleTickerProviderStateMixin we only use this class when we want to use AnimationController so please remove it

File: lib/src/widget/built_in/simple/simple.dart [hunk 4]
```diff
@@ -44,28 +46,42 @@ class SimpleToastWidget extends StatelessWidget {
 
   final bool applyBlurEffect;
 
+  final bool? showCloseButton;
```
Comment (payam-zahedi) [inline lib/src/widget/built_in/simple/simple.dart:49]: Make this parameter non-nullable 

We alwase send a boolean value to the widget

File: lib/src/widget/built_in/simple/simple.dart [hunk 5]
```diff
@@ -77,32 +93,43 @@ class SimpleToastWidget extends StatelessWidget {
...
     required bool applyBlurEffect,
+    required bool showCloseButton,
   }) {
-    Widget body = Container(
+    Widget body;
+
+    if (showCloseButton) {
+      body = Row(
+        children: [
+          Expanded(
```
Comment (payam-zahedi) [inline lib/src/widget/built_in/simple/simple.dart:103]: If you want to have a widget which allocate size only base on its content you can use Flexible widget 

Check out this example:

``` dart 

    Row(
      mainAxisSize: MainAxisSize.min,
      children: [
        Flexible(
          fit: FlexFit.loose,
          child: widget.title ?? const SizedBox(),
        ),
        ToastCloseButton(
          showCloseButton: showCloseButton,
          onCloseTap: widget.onCloseTap,
          icon: widget.defaultStyle.closeIcon(context),
          iconColor: widget.foregroundColor?.withOpacity(.3) ??
              widget.defaultStyle.closeIconColor(context),
        ),
      ],
    )
```

File: lib/src/widget/built_in/simple/simple.dart [hunk 6]
```diff
@@ -77,32 +93,43 @@ class SimpleToastWidget extends StatelessWidget {
...
+          ToastCloseButton(
+            showCloseButton: showCloseButton,
+            onCloseTap: widget.onCloseTap,
+            icon: widget.defaultStyle.closeIcon(context),
+            iconColor: widget.foregroundColor?.withOpacity(.3) ??
+                widget.defaultStyle.closeIconColor(context),
+          ),
+        ],
+      );
+    } else {
```
Comment (payam-zahedi) [inline lib/src/widget/built_in/simple/simple.dart:115]: We don't need this else here becase `ToastCloseButton` will handle its own visibility

**Review by payam-zahedi — CHANGES_REQUESTED** [review]

**Review by AntonioAEMartins — COMMENTED** [review]
@payam-zahedi Thank you so much for your valuable feedback! I’ve implemented all the suggested changes, and now I believe the `simple.dart` file aligns more closely with the overall coding style of the other classes within Toastification.

One additional modification I made was adding a conditional check for `if(showCloseButton)` in the following section:

```dart
Row(
  mainAxisSize: MainAxisSize.min,
  children: [
    Flexible(
      fit: FlexFit.loose,
      child: title ?? const SizedBox(),
    ),
    if (showCloseButton) ...[
      const SizedBox(width: 8),
      ToastCloseButton(
        showCloseButton: showCloseButton,
        onCloseTap: onCloseTap,
        icon: defaultStyle.closeIcon(context),
        iconColor: foregroundColor?.withOpacity(.3) ??
            defaultStyle.closeIconColor(context),
      ),
    ]
  ],
);
```

This conditional ensures that when the `ToastCloseButton` is not displayed, it doesn’t take up unnecessary space, allowing for consistent padding on both the left and right sides of the toast.

Looking forward to your thoughts on these changes!

**Comment by payam-zahedi** [comment]
Thank you, @AntonioAEMartins, for your contributions!

Your changes look solid, and I appreciate the work you put in. If it’s alright with you, I’d like to make a few adjustments to further refine the on-hover close button feature. I believe these tweaks could enhance its functionality and user experience even further.

Let me know if you have any thoughts on this approach!

**Comment by AntonioAEMartins** [comment]
Great to hear @payam-zahedi 

What are the changes that you have in mind?

**Comment by payam-zahedi** [comment]
Currently, the behavior of the close button on hover is somewhat inconsistent. It would improve the user experience if we could add a subtle fade-in/out effect along with a slight shrink in size on hover. This would make the interaction smoother and more visually appealing.

**Comment by payam-zahedi** [comment]
Checkout this video, it seems like the close button’s behavior on appearance and disappearance is a bit abrupt. Adding a smooth fade-in/out animation would enhance the user experience and make the transitions feel more natural.


I do belive we need to add some changes to the CloseButton to make it happend

https://github.com/user-attachments/assets/5d26e5f1-b5b5-4c12-9917-39f71c158caa

**Comment by payam-zahedi** [comment]
This behavior would be a nice improvement!

https://github.com/user-attachments/assets/405eb2d2-ffc1-4710-a0ae-0c0f0290a0d6

I’d suggest achieving it by updating the CloseButton widget, so the change will apply to other Toasts as well. Using an `AnimatedSwitcher` widget could be a good way to implement it.

Here’s a quick example:



``` dart 
import 'package:flutter/material.dart';

/// Flutter code sample for [AnimatedSwitcher].

void main() => runApp(const AnimatedSwitcherExampleApp());

class AnimatedSwitcherExampleApp extends StatelessWidget {
  const AnimatedSwitcherExampleApp({super.key});

  @override
  Widget build(BuildContext context) {
    return const MaterialApp(
      home: AnimatedSwitcherExample(),
    );
  }
}

class AnimatedSwitcherExample extends StatefulWidget {
  const AnimatedSwitcherExample({super.key});

  @override
  State<AnimatedSwitcherExample> createState() =>
      _AnimatedSwitcherExampleState();
}

class _AnimatedSwitcherExampleState extends State<AnimatedSwitcherExample> {
  bool show = true;

  @override
  Widget build(BuildContext context) {
    return ColoredBox(
      color: Colors.white,
      child: Column(
        mainAxisAlignment: MainAxisAlignment.center,
        children: <Widget>[
          Container(
            padding: EdgeInsets.all(8),
            decoration: BoxDecoration(
              color: Colors.blue[100],
              borderRadius: BorderRadius.circular(8),
            ),
            child: Row(
              mainAxisSize: MainAxisSize.min,
              children: [
                Text('Some Text'),
                AnimatedSwitcher(
                  duration: const Duration(milliseconds: 500),
                  transitionBuilder:
                      (Widget child, Animation<double> animation) {
                    return FadeTransition(
                      opacity: animation,
                      child: SizeTransition(
                        sizeFactor: animation,
                        axis: Axis.horizontal,
                        child: child,
                      ),
                    );
                  },
                  child: show
                      ? Container(
                          key: ValueKey(1),
                          padding: const EdgeInsets.all(8),
                          decoration: BoxDecoration(
                            color: Colors.red[200],
                            borderRadius: BorderRadius.circular(8),
                          ),
                          child: Icon(Icons.close),
                        )
                      : SizedBox.shrink(key: ValueKey(2)),
                ),
              ],
            ),
          ),
          ElevatedButton(
            child: const Text('Increment'),
            onPressed: () {
              setState(() {
                show = !show;
              });
            },
          ),
        ],
      ),
    );
  }
}

```

**Comment by AntonioAEMartins** [comment]
Hi @payam-zahedi,

I’ve made some adjustments to the behavior of the Simple Toast Type based on the example code you provided. Overall, the behavior seems correct, but I noticed a slightly odd (though acceptable) behavior when the Simple Toast has multiple lines of text.

Another issue arises in the "Toast Style Selection" menu, where the Simple Toast is occupying all the available vertical space. There are two potential solutions:

Screenshot 2024-10-30 at 14 12 50

The first is to use a Column, but this approach results in incorrect height behavior. When the user hovers over the Toast, the text slightly shifts downward.

The second solution, which resolves the issue entirely, is to use IntrinsicHeight. However, I’m aware that IntrinsicHeight is known to be computationally expensive.

Screenshot 2024-10-30 at 14 14 16

(With IntrinsicHeight)

Do you have any suggestions on how to best address this behavior?

**Comment by payam-zahedi** [comment]
Hi @AntonioAEMartins,

I reviewed your code and made some adjustments to address the issue you encountered. You can see these changes in my latest commit.

Please take a look and let me know if it aligns with your expectations.

**Comment by AntonioAEMartins** [comment]
@payam-zahedi I reviewed the changes, and I really liked the animations and overall behavior. The adjustments align perfectly with what I was hoping for.

**Comment by payam-zahedi** [comment]
@AntonioAEMartins Awesome 

So I will try to resolve conflicts and Merge This PR

**Review by payam-zahedi — APPROVED** [review]
It looks like this PR is now complete.

Thank you very much for your dedication and valuable contributions, @AntonioAEMartins!

**Pull request merged**